File: Algorithms/p144_Binary_Tree_Preorder_Traversal/p144_Binary_Tree_Preorder_Traversal_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        vector<TreeNode*> stack;
        while(1){
            if(root){
                res.push_back(root->val);
                stack.push_back(root->right);
                root = root->left;
            }else{
                if(!stack.empty()){
                    root = stack.back();
                    stack.pop_back();
                }else{
                    break;
                }
            }
        }
        return res;
    }
};

File: Algorithms/p144_Binary_Tree_Preorder_Traversal/p144_Binary_Tree_Preorder_Traversal_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        while True:
            if root:
                res.append(root.val)
                stack.append(root.right)
                root = root.left
            else:
                if stack:
                    root = stack.pop()
                else:
                    break
        return res
